src/Components/test/E2ETest/Tests/FormsTest.cs

```diff
@@ -316,6 +316,27 @@ public void ValidationMessageDisplaysMessagesForField()
             WaitAssert.Empty(emailMessagesAccessor);
         }
 
+        [Fact]
+        public async Task InputComponentsCauseContainerToRerenderOnChange()
+        {
+            var appElement = MountTestComponent<TypicalValidationComponent>();
+            var ticketClassInput = new SelectElement(appElement.FindElement(By.ClassName("ticket-class")).FindElement(By.TagName("select")));
+            var selectedTicketClassDisplay = appElement.FindElement(By.Id("selected-ticket-class"));
+            var select = ticketClassInput.WrappedElement;
+
+            // Shows initial state
+            WaitAssert.Equal("Economy", () => selectedTicketClassDisplay.Text);
+
+            // Refreshes on edit
+            ticketClassInput.SelectByValue("Premium");
+            WaitAssert.Equal("Premium", () => selectedTicketClassDisplay.Text);
+
+            // Leaves previous value unchanged if new entry is unparseable
+            ticketClassInput.SelectByText("(select)");
+            await Task.Delay(500); // Not expecting any UI change, so wait a moment to see if one happens
```

Comment:
No better way to test this?

Reply:
The inherent asynchrony of the server-side execution model makes it tricky for tests to assert "something shouldn't change". It's easy to assert when something should change within a given time span - that's what the `WaitAssert` calls do - but proving a negative is harder.

We *could* add some extra complications to the code under test for cases like this, whereby we try to trigger a secondary effect that we can wait for before running the "nothing changed" assertion. But it's a very unnatural way to write test code.

There are few cases where we really need "nothing changed" assertions in the Components E2E tests, so I haven't regarded it as a major issue. The same test cases also run automatically in the client-side (Blazor) runtime which is fully synchronous, so the assertion is guaranteed to be applied to the post-updates state anyway, meaning that we'd almost definitely catch any cases where the asynchony was hiding a bug.

If we later decide that we really need to solve this in a better way and not rely on waiting like this, I'd prefer to do it once in some central place, by having the test app somehow expose information about the number of renders, so the test code can wait until that number increments.

Reply:
The E2E tests have a bunch of *wait until* loops and so on. 

In general waiting for someone not to happen is just problematic since it will fail to validate what you want if slow.

Reply:
You could add validation and look for the `invalid` class. Not sure if that's a good fit or not for what's being tested.

Reply:
I guess you have validation. So you could wait for the select to have `invalid` and then look at `selectedTicketClassDisplay`

Reply:
Great point. Done.